python/tvm/topi/cuda/conv2d_transpose.py

```diff
@@ -171,124 +182,361 @@ def _fallback_schedule(N, F, Y, X):
-            s[output].bind(bf, te.thread_axis("blockIdx.y"))
-            s[output].bind(s[output].fuse(by, bx), te.thread_axis("blockIdx.x"))
-            s[output].bind(vn, te.thread_axis("vthread"))
-            s[output].bind(vf, te.thread_axis("vthread"))
-            s[output].bind(vy, te.thread_axis("vthread"))
-            s[output].bind(vx, te.thread_axis("vthread"))
-
-            cfg.define_knob("fuse_yx", [0, 1])  # fuse ty,tx or tn,tf
-
-            if cfg["fuse_yx"].val:
-                s[output].bind(tn, te.thread_axis("threadIdx.z"))
-                s[output].bind(tf, te.thread_axis("threadIdx.y"))
-                tyx = s[output].fuse(ty, tx)
-                s[output].bind(s[output].fuse(ty, tx), te.thread_axis("threadIdx.x"))
-                s[OL].compute_at(s[output], tyx)
-
-                # number of threads
-                n_tz = cfg["tile_n"].size[2]
-                n_ty = cfg["tile_f"].size[2]
-                n_tx = cfg["tile_y"].size[2] * cfg["tile_x"].size[2]
-            else:
-                s[output].bind(s[output].fuse(tn, tf), te.thread_axis("threadIdx.z"))
-                s[output].bind(ty, te.thread_axis("threadIdx.y"))
-                s[output].bind(tx, te.thread_axis("threadIdx.x"))
-                s[OL].compute_at(s[output], tx)
-
-                # number of threads
-                n_tz = cfg["tile_n"].size[2] * cfg["tile_f"].size[2]
-                n_ty = cfg["tile_y"].size[2]
-                n_tx = cfg["tile_x"].size[2]
-
-            # tile reduction axes
-            n, f, y, x = s[OL].op.axis
-            rc, ry, rx = s[OL].op.reduce_axis
-            rco, rcm, rci = cfg["tile_rc"].apply(s, OL, rc)
-            s[OL].reorder(rco, rcm, ry, rx, rci, n, f, y, x)
-
-            s[AA].compute_at(s[OL], rx)
-            s[WW].compute_at(s[OL], rx)
-
-            # cooperative fetching
-            for load in [AA, WW]:
-                n, f, y, x = s[load].op.axis
-                fused = s[load].fuse(f, y, x)
-                tz, fused = s[load].split(fused, nparts=n_tz)
-                ty, fused = s[load].split(fused, nparts=n_ty)
-                tx, fused = s[load].split(fused, nparts=n_tx)
-                s[load].bind(tz, te.thread_axis("threadIdx.z"))
-                s[load].bind(ty, te.thread_axis("threadIdx.y"))
-                s[load].bind(tx, te.thread_axis("threadIdx.x"))
-
-            s[output].pragma(kernel_scope, "auto_unroll_max_step", cfg["auto_unroll_max_step"].val)
-            s[output].pragma(kernel_scope, "unroll_explicit", cfg["unroll_explicit"].val)
+        if op.tag == "group_conv2d_transpose_nchw":
+            _schedule_group_conv2d_transpose_nchw(cfg, s, op)
+
+        elif op.tag == "conv2d_transpose_nchw":  # groups == 1 cases delegate to regular conv2d_transpose
+            _schedule_conv2d_transpose_nchw(cfg, s, op)
 
     traverse_inline(s, outs[0].op, _callback)
 
     return s
 
 
+def _schedule_group_conv2d_transpose_nchw(cfg, s, op):
```

Comment:
Why do we need this schedule? Can we reuse the group = 1 one?

Reply:
`_schedule_conv2d_transpose_nchw` gives me the following error when groups > 1
``` 
RuntimeError: Memory verification failed with the following errors:
Did you forget to bind?
    Variable `W` is directly accessed by host memory (it is not contained in a thread environment or in the function arguments.
```

`_schedule_group_conv2d_transpose_nchw` has the following lines which I believe alleviate that:
```python
    ...
    s[kernel_transform].compute_inline()

    s[kernel_transform].set_scope("shared")
    WW = kernel_transform
    ...
```

Also the tiling in `_schedule_group_conv2d_transpose_nchw` is different, similar to the group_conv2d_schedule:
```python
    cfg.define_split("tile_n", cfg.axis(n), num_outputs=4)
    cfg.define_split("tile_g", cfg.axis(groups), num_outputs=2)
    cfg.define_split("tile_f", cfg.axis(num_filters // groups), num_outputs=4)
    cfg.define_split("tile_y", cfg.axis(y), num_outputs=4)
    cfg.define_split("tile_x", cfg.axis(x), num_outputs=4)
    cfg.define_split("tile_rc", cfg.axis(rc), num_outputs=3)
```
vs `_schedule_conv2d_transpose_nchw`
```python
    cfg.define_split("tile_n", cfg.axis(n_tuning_axis), num_outputs=4)
    cfg.define_split("tile_f", cfg.axis(f), num_outputs=4)
    cfg.define_split("tile_y", cfg.axis(y), num_outputs=4)
    cfg.define_split("tile_x", cfg.axis(x), num_outputs=4)
    cfg.define_split("tile_rc", cfg.axis(rc), num_outputs=3)
```
This causes some more changes with reorder/binding later on.

I could try merging them to one function with those changes conditioned on `groups` if you like.

Reply:
OK, I took another look at it and simplified it a lot.

I realized that conv2d_transpose_nchw was already doing the `kernel_transform` thing inline in te.compute and that all the indexing math in that call simplified correctly even with groups=1 (duhh :sweat_smile:).

So now the original schedule function works fine (although it won't tile over groups like group_conv2d, but I guess all these handwritten schedules are getting phased out eventually anyways).

Reply:
> although it won't tile over groups like group_conv2d

You can add such thing to the existing schedule. You can change the tag name at https://github.com/apache/tvm/blob/66f23e2afb31e88c7616b280423e0aeab5bce0e4/python/tvm/topi/cuda/conv2d_transpose.py#L126 when group > 1, and inside the existing schedule, you check the tag name to decide if you enable tiling over groups.